File: python_scripts/pomodoro.py
import sys, time
from getch import getch

def focus(minutes):
    seconds = 60
    for i in range(minutes*60-1, 0, -1):
        if i <= 0:
            break
        else:
            seconds -= 1
            mins =  int(i/60)
            if mins < 10 and seconds == 0:
                sys.stdout.write("\rfocus...  {}:00".format(int(mins)))
                sys.stdout.flush()
                seconds = 60
            elif mins < 10 and seconds > 0:
                if seconds < 10:
                    sys.stdout.write("\rfocus...  {}:0{}".format(int(mins), seconds))
                    sys.stdout.flush()
                else:
                    sys.stdout.write("\rfocus...  {}:{}".format(int(mins), seconds))
                    sys.stdout.flush()
            elif mins >= 10 and seconds == 0:
                sys.stdout.write("\rfocus... {}:00".format(int(mins)))
                sys.stdout.flush()
                seconds = 60
            else:
                if seconds < 10:
                    sys.stdout.write("\rfocus... {}:0{}".format(int(mins), seconds))
                    sys.stdout.flush()
                else:
                    sys.stdout.write("\rfocus... {}:{}".format(int(mins), seconds))
                    sys.stdout.flush()
            time.sleep(1)
def rest(minutes):
    seconds = 60
    for i in range(minutes*60-1, 0, -1):
        if i <= 0:
            break
        else:
            seconds -= 1
            mins =  int(i/60)
            if mins < 10 and seconds == 0:
                sys.stdout.write("\rrest...   {}:00".format(int(mins)))
                sys.stdout.flush()
                seconds = 60
            elif mins < 10 and seconds > 0:
                if seconds < 10:
                    sys.stdout.write("\rrest...   {}:0{}".format(int(mins), seconds))
                    sys.stdout.flush()
                else:
                    sys.stdout.write("\rrest...   {}:{}".format(int(mins), seconds))
                    sys.stdout.flush()
            elif mins >= 10 and seconds == 0:
                sys.stdout.write("\rrest...  {}:00".format(int(mins)))
                sys.stdout.flush()
                seconds = 60
            else:
                if seconds < 10:
                    sys.stdout.write("\rrest...  {}:0{}".format(int(mins), seconds))
                    sys.stdout.flush()
                else:
                    sys.stdout.write("\rrest...  {}:{}".format(int(mins), seconds))
                    sys.stdout.flush()
            time.sleep(1)

def pomodoro(cycles):
    while cycles != 0:
        focus(25)
        rest(5)
        cycles -= 1
        sys.stdout.write("\rrest...  pomodoro cycle complete\n")
        sys.stdout.flush()

pomodoro(4)

File: rust_scripts/pomodoro.rs
use std::{io::{stdout, Write}, thread::sleep, time::Duration};
use clap::Parser;

#[derive(Parser, Default, Debug)]
#[clap(author="angelshatepop", version, about="pomodoro clock")]
struct Args {
   focus: u32,
   rest: u32,
   pomodoros: u32,
}

pub fn pomodoro_timer(title: &str, mins: u32, state: bool) {
    let mut stdout = stdout();
    let (mut seconds, mut seconds_total) = (60, mins*60);
    loop {
        if seconds_total == 0 {
            break;
        } else {
            seconds -= 1;
            seconds_total -= 1;
            let mins = seconds_total / 60;
            let mut time_str = format!("\r{}... {:02}:{:02}", title, mins, seconds);
            if state == false {
                time_str = format!("\r{}...  {:02}:{:02}", title, mins, seconds);
            }
            stdout.write(time_str.as_bytes()).unwrap();
            stdout.flush().unwrap();
            if seconds == 0 {seconds = 60};
            sleep(Duration::from_secs(1));
        }
    }
}

pub fn pomodoro(focus_len: u32, rest_len: u32, cycles: u32) {
    print!("{esc}c", esc = 27 as char);
    let mut cycle = 0;
    while cycle != cycles {
        pomodoro_timer("focus", focus_len, true);
        pomodoro_timer("rest", rest_len, false);
        println!("  --pomodoro({}) complete", cycle+1);
        cycle += 1;
    }
    println!("\npomodoro session complete!")
}

fn main () {
    let args = Args::parse();
    pomodoro(args.focus, args.rest, args.pomodoros)
}